### PR TITLE
Uart mode tests

### PR DESCRIPTION
**Pull request by yegorich**
Add support for new UART API call `uart_mode`.

Before performing these new tests, the DUT will be checked whether `uart_mode` is supported. If not, the test will be marked as **PASS** with a note "Feature is not supported".

### CODE REVIEW

File: tests/periph_uart/tests/02__periph_uart_mode.robot [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+Resource            periph_uart.keywords.txt
+Resource            api_shell.keywords.txt
+
+Variables           test_vars.py
+
+Force Tags          periph  uart
+
+*** Test Cases ***
+Even Parity 8 Bits
+    ${status}   ${value}=       Run Keyword And Ignore Error   API Call Should Succeed   Uart Mode   data_bits=8   parity="N"   stop_bits=1
```
Comment (smlng) [inline tests/periph_uart/tests/02__periph_uart_mode.robot]: looks like these two lines are used in all tests here and should be wrapped into a "speaking" keyword?
Something like `DUT Should Have UART Mode` maybe (@MrKevinWeiss what do you think?)

Comment (MrKevinWeiss) [reply tests/periph_uart/tests/02__periph_uart_mode.robot]: Maybe `UART mode should exist` or `DUT UART mode should exist`

**Comment by smlng** [comment]
@MrKevinWeiss does this (`uart.ctrl.data_bits`)  need the new firmware and RIOT-PAL version?

**Comment by yegorich** [comment]
@smlng yes, you need the latest PHILIP fw and the newest memory map.

**Comment by MrKevinWeiss** [comment]
So maybe we wait until the philip update is done and everything is updated before we merge it in.

**Comment by yegorich** [comment]
v2 changes:

- add new keyword: "DUT UART mode should exist"
- rework "DUT Should Match String"  to accept a reference string
- add 38400 and 9600 baud rate tests

File: tests/periph_uart/tests/02__periph_uart_mode.robot (hunk 1 above)
Comment (yegorich) [reply tests/periph_uart/tests/02__periph_uart_mode.robot]: I've used "DUT UART mode should exist"

**Comment by MrKevinWeiss** [comment]
So I am trying to update to the new philip and riot pal... I caught a few bugs on philip but one thing I found is it appears that the nucleo-l476re mode setting to 8E1 actually sets it to 6E1.  On the scope it can decide at these settings at least.  I do get a framing error in PHiLIP though.  I'll pass along my changes but it is still using devel level philip and unreleased riot pal.  So for information only I guess.

**Comment by MrKevinWeiss** [comment]
[PHiLIP](https://github.com/MrKevinWeiss/PHiLIP/commit/b3808f8a3302e50ad565269e541d5085415496ed)
[RIOT_pal](https://github.com/MrKevinWeiss/riot_pal/commit/76ea701b2e915d2f875912e6c1ce83577d749f9d)
[RF](https://github.com/MrKevinWeiss/RobotFW-tests/commit/a0990bdfc58a8a13b612013abd25d65142688ec6)

**Comment by MrKevinWeiss** [comment]
It seems to pass on the nucleo-f401re, everything but the two stop bits test.   Maybe it is just the implementation in riot.

**Comment by yegorich** [comment]
v3 changes:

* add PHiLIP statistics to both passed and failed tests

```
------------------------------------------------------------------------------
Wrong Baud Test                                                       | PASS |
RX: 0, TX: 0, PE: 0, FE: 1, NF: 0, ORE: 0                                    
------------------------------------------------------------------------------
Baud Test 38400                                                       | PASS |
RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0                                    
------------------------------------------------------------------------------
Extended Echo                                                         | FAIL |
RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0: u222 != t111
------------------------------------------------------------------------------
Echo                                                                  | FAIL |
RX: 0, TX: 0, PE: 0, FE: 1, NF: 0, ORE: 0:  != t111                          
```

**Comment by yegorich** [comment]
v4 changes:

* rework `DUT Should Match String` and `DUT Should Not Match String` keywords. Both keywords use `PeriphUartUtil` class to create an error message based on `Uart Send String` result. Error message consists of error cause and PHiLIP statistics.

```
------------------------------------------------------------------------------
Echo                                                                  | FAIL |                                       
Reference string doesn't match the received one: RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Long Echo                                                             | FAIL |                                       
Test timed out: RX: 0, TX: 0, PE: 0, FE: 0, NF: 0, ORE: 0           
------------------------------------------------------------------------------
Wrong Baud Test                                                       | FAIL |                                       
No timeout occurred and the reference string does match the received one: RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0 
------------------------------------------------------------------------------
```

**Comment by yegorich** [comment]
Rebased and tested against Nucleo-f103rb.

**Comment by smlng** [comment]
@MrKevinWeiss can you test this, too? Would be a good addition since we updated the HIL infrastructure to support this.

**Comment by MrKevinWeiss** [comment]
Test results
```xml
<testsuite name="tests_periph_uart" tests="14" errors="0" failures="0" skipped="0" time="117.990"><testcase classname="tests_periph_uart.Periph Uart Base" name="Echo" time="1.065">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Long Echo" time="1.080">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Extended Echo" time="1.064">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Extended Long Echo" time="1.083">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Register Access" time="1.042">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Should Not Access Invalid Register" time="1.032">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Baud Test 38400" time="1.066">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Baud Test 9600" time="1.068">
</testcase><testcase classname="tests_periph_uart.Periph Uart Base" name="Wrong Baud Test" time="21.107">
</testcase><testcase classname="tests_periph_uart.Periph Uart Mode" name="Even Parity 8 Bits" time="21.238">
</testcase><testcase classname="tests_periph_uart.Periph Uart Mode" name="Odd Parity 8 Bits" time="21.252">
</testcase><testcase classname="tests_periph_uart.Periph Uart Mode" name="Even Parity 7 Bits" time="21.264">
</testcase><testcase classname="tests_periph_uart.Periph Uart Mode" name="Odd Parity 7 Bits" time="21.266">
</testcase><testcase classname="tests_periph_uart.Periph Uart Mode" name="Two Stop Bits" time="1.177">
</testcase></testsuite>
```


terminal
```
==============================================================================
tests_periph_uart                                                             
==============================================================================
tests_periph_uart.Periph Uart Base                                            
==============================================================================
Echo                                                                  | PASS |
RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Long Echo                                                             | PASS |
RX: 44, TX: 44, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Extended Echo                                                         | PASS |
RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Extended Long Echo                                                    | PASS |
RX: 44, TX: 44, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Register Access                                                       | PASS |
RX: 8, TX: 43, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Should Not Access Invalid Register                                    | PASS |
RX: 9, TX: 14, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Baud Test 38400                                                       | PASS |
RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Baud Test 9600                                                        | PASS |
RX: 5, TX: 5, PE: 0, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Wrong Baud Test                                                       | PASS |
RX: 0, TX: 0, PE: 0, FE: 1, NF: 0, ORE: 0
------------------------------------------------------------------------------
tests_periph_uart.Periph Uart Base                                    | PASS |
9 critical tests, 9 passed, 0 failed
9 tests total, 9 passed, 0 failed
==============================================================================
tests_periph_uart.Periph Uart Mode                                            
==============================================================================
Even Parity 8 Bits                                                    | PASS |
RX: 5, TX: 5, PE: 1, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Odd Parity 8 Bits                                                     | PASS |
RX: 5, TX: 5, PE: 1, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Even Parity 7 Bits                                                    | PASS |
RX: 5, TX: 5, PE: 1, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Odd Parity 7 Bits                                                     | PASS |
RX: 5, TX: 5, PE: 1, FE: 0, NF: 0, ORE: 0
------------------------------------------------------------------------------
Two Stop Bits                                                         | PASS |
RX: 6, TX: 6, PE: 0, FE: 1, NF: 0, ORE: 0
------------------------------------------------------------------------------
tests_periph_uart.Periph Uart Mode                                    | PASS |
5 critical tests, 5 passed, 0 failed
5 tests total, 5 passed, 0 failed
==============================================================================
tests_periph_uart                                                     | PASS |
14 critical tests, 14 passed, 0 failed
14 tests total, 14 passed, 0 failed
==============================================================================

```

**Comment by MrKevinWeiss** [comment]
@smlng is everything OK on your end?

**Pull request merged**